ws-client/src/client.rs

```diff
@@ -375,8 +420,50 @@
 					}
 				}
 			}
-			Either::Right((Some(Ok(jsonrpc::Response::Batch(_responses))), _)) => {
-				log::warn!("Ignore batch response not supported, #103");
+			Either::Right((Some(Ok(jsonrpc::Response::Batch(batch))), _)) => {
+				let mut digest = BTreeSet::new();
+				let mut rps_unordered: Vec<_> = Vec::new();
+
+				for rp in batch {
+					let id = match rp.id().as_number().copied() {
+						Some(id) => id,
+						None => {
+							let _ = front_error.send(Error::InvalidRequestId);
+							return;
+						}
+					};
+					let rp: Result<JsonValue, Error> = rp.try_into().map_err(Error::Request);
+					let rp = match rp {
+						Ok(rp) => rp,
+						Err(err) => {
+							let _ = front_error.send(err);
+							return;
+						}
+					};
+					digest.insert(id);
+					rps_unordered.push((id, rp));
+				}
+
+				let (batch_ordered, send_back) = match manager.complete_pending_batch(digest) {
+					Some((batch, send_back)) => (batch, send_back),
+					None => {
+						log::warn!("Received unknown batch response");
+						continue;
+					}
+				};
+
+				let mut ordered_responses = Vec::new();
+				for id in batch_ordered {
+					// NOTE(niklasad1): O(n)
+					let pos = match rps_unordered.iter().position(|(i, _)| *i == id) {
+						Some(pos) => pos,
+						None => unreachable!("All request ID's valid checked by RequestManager above; qed"),
+					};
+					// Dummy value should never be.
+					let (_, rp) = std::mem::replace(&mut rps_unordered[pos], (0, JsonValue::Number(0.into())));
```

Comment:
Is there a guarantee that `Id` is non-zero? I took a glance and it seems that it boils down to a `u64` so using 0 as the tombstone is perhaps not 100% safe?

Reply:
This is just dummy value because each request IDs should be unique and read exactly once, ensured by the RequestManager if we don't have any bugs :)
 
So think of this as `std::mem::take` but I could instead clone the value if you like it better.

Reply:
`std::mem::replace` is fine. My concern was more that we could end up with a valid ID of `0` and that might be bad? We are using `0` to signal "free" and the comment above, "Dummy value should never be", made me think `0` must never be used for an ID. If this is the case then we should use the type system to ensure IDs are non-zero.

Reply:
Switched to `Vec::swap_remove` it is `O(1)` but doesn't preserve order.